### PR TITLE
[API-537] Add support for the binary metrics

### PR DESCRIPTION
**Pull request by mdumandag**
Implemented the mechanism to compress the metrics we collect in a
way that is compatible with the server-side.

With this change, the latest Management Center versions are able to
show statistics regarding the Python client.

For backward compatibility, we also continue sending metrics in
a comma separated key-value format (client attributes).

In this PR, I also significantly cleaned up the statistics service.
In the old code, we were trying to reuse the results of some
calls. In the benchmarks I did, I saw that it didn't really improve
much. (New approach approx. 0.5 ms, old approach approx. 0.4 ms per
statistics collection) This really small difference does not justify
the need for complexity so I went with the simpler approach.

Also, aligned the log levels with the Java client for the various
warnings.

Also, bumped the version of psutil to 5.8.0.

### CODE REVIEW

**Comment by mdumandag** [comment]
Here is the benchmark I used to compare the old vs new approach
```python
import timeit
import psutil


def f1():
    psutil.virtual_memory().total
    psutil.virtual_memory().free
    psutil.virtual_memory().used
    psutil.swap_memory().total
    psutil.swap_memory().free
    psutil.cpu_percent()
    psutil.cpu_count()
    p = psutil.Process()
    p.memory_info().rss
    p.num_fds()
    p.rlimit(psutil.RLIMIT_NOFILE)[1]
    p.cpu_times()
    p.create_time()


def f2():
    v = psutil.virtual_memory()
    v.total
    v.free
    v.used
    s = psutil.swap_memory()
    s.total
    s.free
    psutil.cpu_percent()
    psutil.cpu_count()
    p = psutil.Process()
    with p.oneshot():
        p.memory_info().rss
        p.num_fds()
        p.rlimit(psutil.RLIMIT_NOFILE)[1]
        p.cpu_times()
        p.create_time()


print(timeit.timeit(f1, number=10000))
print(timeit.timeit(f2, number=10000))
```

File: tests/unit/metrics/metrics_test.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            MetricDescriptor(
+                metric="longPrefixMetric",
+                prefix="a" * 254,
+                unit=ProbeUnit.BYTES,
+            ),
+            2147483647,
+        )
+
+        actual_blob = compressor.generate_blob()
+        self.assertEqual(expected_blob, actual_blob)
```
Comment (srknzl) [inline tests/unit/metrics/metrics_test.py:54]: nodejs test has a extractPayload method. Does python and java create exact same binary?

File: hazelcast/metrics.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    +--------------------------------+--------------------+
+    | Metrics blob                   |   variable size    |
+    +--------------------------------+--------------------+
+    """
+
+    __slots__ = ("_metrics_buf", "_dict_buf", "_metrics_dict", "_metrics_count", "_last_descriptor")
+
+    def __init__(self):
+        self._metrics_buf = _OutputBuffer()
+        self._dict_buf = _OutputBuffer()
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/metrics.py:156]: I tried to follow Node.js for the default values. Java might have a higher number as there are almost 5x stats collected in the Java client, compared to Node.js and Python

File: hazelcast/statistics.py [hunk 3]
```diff
@@ -36,18 +38,23 @@ def __init__(
...
         def _statistics_task():
             if not self._client.lifecycle_service.is_running():
                 return
 
-            self._send_statistics()
-            self._statistics_timer = self._reactor.add_timer(self._period, _statistics_task)
+            try:
+                self._collect_and_send_stats()
+            finally:
+                self._statistics_timer = self._reactor.add_timer(self._period, _statistics_task)
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/statistics.py:57]: No, this just adds something to an unbounded queue

File: tests/unit/metrics/metrics_test.py (hunk 1 above)
Comment (mdumandag) [reply tests/unit/metrics/metrics_test.py:54]: Yes, they produce the same binary. I tested this on Linux and Windows as well

File: hazelcast/statistics.py [hunk 4]
```diff
@@ -57,240 +64,309 @@ def shutdown(self):
...
+            "os.totalSwapSpaceSize",
+            lambda: psutil.swap_memory().total,
+        )
+        self._register_system_gauge(
+            "os.freeSwapSpaceSize",
+            lambda: psutil.swap_memory().free,
+        )
+        self._register_system_gauge(
+            "os.systemLoadAverage",
+            lambda: psutil.cpu_percent(),
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/statistics.py:97]: Good catch. I changed this with https://docs.python.org/3/library/os.html#os.getloadavg

File: hazelcast/statistics.py [hunk 5]
```diff
@@ -57,240 +64,309 @@ def shutdown(self):
...
+            lambda p: to_nanos(sum(p.cpu_times())),
+        )
+        self._register_process_gauge(
+            "runtime.uptime",
+            lambda p: to_millis(current_time() - p.create_time()),
+        )
+
+    def _register_system_gauge(self, gauge_name, gauge_fn, value_type=ValueType.LONG):
+        try:
+            gauge_fn()
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/statistics.py:129]: added the comment to both functions

**Comment by srknzl** [comment]
Before merging, can you test that if metrics can be sent to management center on windows(python client runs on windows)? some of them like number of file descriptors may not work.

File: hazelcast/statistics.py [hunk 6]
```diff
@@ -1,30 +1,32 @@
...
+_ATTRIBUTE_SEPARATOR = ","
+_KEY_VALUE_SEPARATOR = "="
+_EMPTY_ATTRIBUTE_VALUE = ""
 
-class Statistics(object):
-    _NEAR_CACHE_CATEGORY_PREFIX = "nc."
-    _STAT_SEPARATOR = ","
-    _KEY_VALUE_SEPARATOR = "="
-    _EMPTY_STAT_VALUE = ""
+_DEFAULT_PROBE_VALUE = 0
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/statistics.py]: no, it was something leftover. removed it

File: hazelcast/metrics.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+    | Compressor version             |   2 bytes (short)  |
+    +--------------------------------+--------------------+
+    | Size of dictionary blob        |   4 bytes (int)    |
+    +--------------------------------+--------------------+
+    | Dictionary blob                |   variable size    |
+    +--------------------------------+--------------------+
+    | Number of metrics in the blob  |   4 bytes (int)    |
+    +--------------------------------+--------------------+
+    | Metrics blob                   |   variable size    |
+    +--------------------------------+--------------------+
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/metrics.py:149]: Actually, the text above this was one space off-indented :smile: Fixed that

File: hazelcast/metrics.py [hunk 8]
```diff
@@ -213,7 +213,7 @@ def _write_descriptor(self, descriptor):
             else:
                 self._metrics_buf.write_byte(descriptor.unit)
 
-        # include excludedTargets and tags bytes for compatibility purposes
+        # include excludedTargets and tags bits for compatibility purposes
```
Comment (srknzl) [inline hazelcast/metrics.py:216]: This should be bytes by the way

**Comment by mdumandag** [comment]
Yes, I have verified this, on Windows, this feature works with a subset of the stats we sent from Linux.

For example, file descriptor count is not sent on Windows, as there is no such thing in Windows, or the load average will not be sent, as it is not implemented there in the Python standard library.

File: hazelcast/metrics.py (hunk 8 above)
Comment (mdumandag) [reply hazelcast/metrics.py]: Correct, sorry about that.

File: hazelcast/statistics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply hazelcast/statistics.py:57]: Then why you added it? as a precaution maybe?

[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/statistics.py:57]: Oh, I thought you were talking about the `self._reactor.add_timer` call. 

The `_collect_and_send_stats`, might fail in some unexpected ways, like the zlib.compress might throw (or let's say I am not sure). Yes, basically, this is a precaution

**Pull request merged**